haystack_experimental/components/preprocessors/md_header_level_inferrer.py

```diff
@@ -0,0 +1,110 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+
+from haystack import Document, component, logging
+
+
+@component
+class MarkdownHeaderLevelInferrer:
+    """
+    Infers and rewrites header levels in markdown text to normalize hierarchy.
+    """
+
+    def __init__(self):
+        """
+        Initializes the MarkdownHeaderLevelInferrer.
+
+        Uses a hardcoded regex pattern to match markdown headers from level 1 to 6.
+        """
+        self._header_pattern = r"(?m)^(#{1,6}) (.+)$"
```

Comment:
pre-compile the regex pattern in `__init__` for better performance:

```suggestion
        self._header_pattern = re.compile(r"(?m)^(#{1,6}) (.+)$")
```